azure-iot-device/azure/iot/device/common/pipeline/pipeline_ops_base.py

```diff
@@ -44,10 +51,79 @@ def __init__(self, callback):
                 "Cannot instantiate PipelineOperation object.  You need to use a derived class"
             )
         self.name = self.__class__.__name__
-        self.callback = callback
+        self.callbacks = []
         self.needs_connection = False
         self.completed = False
 
+        self.callbacks.append(callback)
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def add_callback(self, callback):
+        self.callbacks.append(callback)
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def complete(self, error=None):
+        if error:
+            logger.error("{}: completing with error {}".format(self.name, error))
+        else:
+            logger.debug("{}: completing without error".format(self.name))
+
+        if self.completed:
+            logger.error("{}: has already been completed!".format(self.name))
+            e = pipeline_exceptions.OperationError(
+                "Attempting to complete an already-completed operation: {}".format(self.name)
+            )
+            # This could happen in a foreground or background thread, so err on the side of caution
+            # and send it to the background handler.
+            handle_exceptions.handle_background_exception(e)
+        else:
+            self.completed = True
+
+            while self.callbacks:
+                if not self.completed:
+                    logger.debug(
+                        "{}: Operation no longer complete! Suspending completion".format(self.name)
+                    )
+                    break
+                callback = self.callbacks.pop()
+                try:
+                    callback(op=self, error=error)
+                except Exception as e:
+                    logger.error(
+                        "Unhandled error while triggering callback for {}".format(self.name)
+                    )
+                    logger.error(traceback.format_exc())
+                    # This could happen in a foreground or background thread, so err on the side of caution
+                    # and send it to the background handler.
+                    handle_exceptions.handle_background_exception(e)
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def uncomplete(self):
+        logger.debug("{}: Uncompleting".format(self.name))
+        self.completed = False
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def spawn_worker_op(self, worker_op_type, **kwargs):
+        logger.debug("{}: creating worker op of type {}".format(self.name, worker_op_type.__name__))
+
+        # CT-TODO: does this need to use a weakref?
```

Comment:
># CT-TODO: does this need to use a weakref? [](start = 8, length = 43)

personal opinion: it doesn't hurt to use weak references, but why make the code more confusing?  I think we only need weak references if we're passing a callback to a timer function or to an external component.  All other cases are (I think) superfluous.  Maybe we need a coding guideline for this in the wiki?

Reply:
Okay good to know, thanks.

---
In reply to: [345994008](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r345994008) [](ancestors = 345994008)